rpc.go

```diff
@@ -12,22 +18,99 @@ func (r *RPCError) Error() string {
 
 // RPCRequest represents an incoming JSON-RPC request
 type RPCRequest struct {
-	ID     interface{}    `json:"id" msgpack:"id"`
+	ID     any    `json:"id" msgpack:"id"`
 	Async  bool   `json:"async,omitempty" msgpack:"async,omitempty"`
 	Method string `json:"method,omitempty" msgpack:"method,omitempty"`
-	Params []interface{}  `json:"params,omitempty" msgpack:"params,omitempty"`
+	Params []any  `json:"params,omitempty" msgpack:"params,omitempty"`
 }
 
 // RPCResponse represents an outgoing JSON-RPC response
 type RPCResponse struct {
-	ID     interface{}       `json:"id" msgpack:"id"`
+	ID     any       `json:"id" msgpack:"id"`
 	Error  *RPCError `json:"error,omitempty" msgpack:"error,omitempty"`
-	Result interface{}       `json:"result,omitempty" msgpack:"result,omitempty"`
+	Result any       `json:"result,omitempty" msgpack:"result,omitempty"`
 }
 
 // RPCNotification represents an outgoing JSON-RPC notification
 type RPCNotification struct {
-	ID     interface{}    `json:"id" msgpack:"id"`
+	ID     any    `json:"id" msgpack:"id"`
 	Method string `json:"method,omitempty" msgpack:"method,omitempty"`
-	Params []interface{}  `json:"params,omitempty" msgpack:"params,omitempty"`
+	Params []any  `json:"params,omitempty" msgpack:"params,omitempty"`
+}
+
+type RPCRawResponse struct {
+	id      string
+	Data    []byte
+	Decoded *RPCResponse
+
+	decodedError bool
+	hasError     bool
+	error        *RPCError
+}
+
+func (res *RPCRawResponse) Error() error {
+	if res.HasError() {
+		return res.error
+	}
+
+	return nil
+}
+
+func (res *RPCRawResponse) HasError() bool {
```

Comment:
This is wayyy too complicated for a Go error check
The way to check errors in go is `if err != nil`, so this work should be done before the user gets the result back

Reply:
Agreed, it needs some refactoring

My Idea was that we run the check once, and then we rely on the error/hasError states from the above suggestion